### DISCUSSION TITLE
在编译linux的demo时报错

### DISCUSSION THREAD

**Discussion by jackTJT** [Tencent/libpag#1939] [General] [answered]
Consolidate compiler generated dependencies of target pag-linux
[ 50%] Linking CXX executable pag-linux
../vendor/libpag/linux/x64/libpag.a(Directory.cpp.o): In function `pag::Directory::CreateRecursively(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
Directory.cpp:(.text+0x7b2): undefined reference to `std::filesystem::__cxx11::path::_M_split_cmpts()'
Directory.cpp:(.text+0x7ba): undefined reference to `std::filesystem::create_directories(std::filesystem::__cxx11::path const&)'
../vendor/libpag/linux/x64/libpag.a(Directory.cpp.o): In function `pag::Directory::VisitFiles(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long)>)':
Directory.cpp:(.text+0x8f0): undefined reference to `std::filesystem::__cxx11::path::_M_split_cmpts()'
Directory.cpp:(.text+0x901): undefined reference to `std::filesystem::__cxx11::directory_iterator::directory_iterator(std::filesystem::__cxx11::path const&, std::filesystem::directory_options, std::error_code*)'
Directory.cpp:(.text+0x9f3): undefined reference to `std::filesystem::__cxx11::directory_iterator::operator*() const'
Directory.cpp:(.text+0xb26): undefined reference to `std::filesystem::file_size(std::filesystem::__cxx11::path const&)'
Directory.cpp:(.text+0xb9d): undefined reference to `std::filesystem::__cxx11::directory_iterator::operator++()'
Directory.cpp:(.text+0xc9c): undefined reference to `std::filesystem::status(std::filesystem::__cxx11::path const&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/pag-linux.dir/build.make:101: pag-linux] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/pag-linux.dir/all] Error 2
gmake: *** [Makefile:91: all] Error 2

**Comment by 0x1306a94** [comment]
@jackTJT gcc 版本是多少。filesystem 需要gcc 8.0+

**Reply by dophinlife** [reply]
GCC 版本 8.3.1，遇到了同样的问题。

> Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/opt/rh/devtoolset-8/root/usr/libexec/gcc/x86_64-redhat-linux/8/lto-wrapper
Target: x86_64-redhat-linux
Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,lto --prefix=/opt/rh/devtoolset-8/root/usr --mandir=/opt/rh/devtoolset-8/root/usr/share/man --infodir=/opt/rh/devtoolset-8/root/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --with-linker-hash-style=gnu --with-default-libstdcxx-abi=gcc4-compatible --enable-plugin --enable-initfini-array --with-isl=/builddir/build/BUILD/gcc-8.3.1-20190311/obj-x86_64-redhat-linux/isl-install --disable-libmpx --enable-gnu-indirect-function --with-tune=generic --with-arch_32=x86-64 --build=x86_64-redhat-linux
Thread model: posix
gcc version 8.3.1 20190311 (Red Hat 8.3.1-3) (GCC)

**Comment by domchen** [comment]
编译PAG需要 C++17的支持，理论上 GCC 8.0 以上就可以了。你可以尝试三个方向：1、升级到更高的 GCC 版本试一下 2、源码也更新到 main 分支最新代码尝试编译 3、清空一下你本地的缓存，特别是 cmake的build目录，全新编译。

**Reply by dophinlife** [reply]
升级后解决了。gcc升级到 11.2.1，libstdc++.so.6 升级到 6.0.26。

**Comment by jackTJT** [comment]
这边问题已经解决，出现std::filesystem报错的原因是因为cmakeList没有对libstdc++fs.a进行配置，不同服务器有不同的默认情况，在有些服务器上是默认不加入libstdc++fs.a的，需要在cmakeList中进行配置

**Comment by 0x1306a94** [comment]
@jackTJT 我之前测试好像 GCC 8.3+ 就不用添加 stdc++fs 依赖了，不过最后我还是用的GCC 9

**Comment by domchen** [comment, marked as answer]
> 这边问题已经解决，出现std::filesystem报错的原因是因为cmakeList没有对libstdc++fs.a进行配置，不同服务器有不同的默认情况，在有些服务器上是默认不加入libstdc++fs.a的，需要在cmakeList中进行配置

查了一下官网文档：https://gcc.gnu.org/gcc-9/changes.html，GCC 9.0+ 才内置了 C++17 的 filesystem 支持。那我们更新一下 Linux 编译要求吧，提高到 GCC 9.0+。
